### PR TITLE
Create azure-cli-core package that command modules and azure-cli can depend on

### LINKED ISSUE #658: Restructure azure-cli package (and command module loading)
The azure-cli package should be a thin entry point to the CLI.
This package will depend on 'azure-cli-core' which holds logic for commands, outputs, etc.
This package can then have dependencies on any command packages that it may need.

Command packages should depend on a package called 'azure-cli-core' to allow command packages to register commands, etc.

This will provide a clear interface between command packages and the core of the CLI.

### PR DESCRIPTION
**Pull request by derekbekoe**
Fixes #658
- Nightly build script modifications
- Load module and fallback to all if doesn't work
- Fix Dockerfile
- No modules found, no problem
- Modify package_verify to work with new package structure
- modify .travis.yml file for CI builds to work again
- fix dev_setup
- import modules differently
- Move stuff

### CODE REVIEW

**Comment by derekbekoe** [comment]
@tjprescott, @yugangw-msft, @burtbiel, @johanste please take a look.

File: scripts/install/install.py [hunk 1]
```diff
@@ -45,17 +44,14 @@
...
-$PATH_TO_PYTHON -m azure.cli "$@"
+{install_dir}/{bin_dir_name}/python -m azure.cli "$@"
 """
 
 DEFAULT_INSTALL_DIR = os.path.join(os.path.sep, 'usr', 'local', 'az')
 DEFAULT_EXEC_DIR = os.path.join(os.path.sep, 'usr', 'local', 'bin')
 VIRTUALENV_VERSION = '15.0.0'
 BIN_DIR_NAME = 'Scripts' if platform.system() == 'Windows' else 'bin'
 EXECUTABLE_NAME = 'az'
-ENVS_DIR_NAME = 'envs'
```
Comment (tjprescott) [inline scripts/install/install.py]: Is this going away?

File: scripts/install/install.py [hunk 2]
```diff
@@ -87,37 +83,31 @@ def create_virtualenv(tmp_dir, version, install_dir):
...
+        param_pre_release=param_pre_release,
     )
 
 def install_cli(install_dir, tmp_dir):
     path_to_pip = os.path.join(install_dir, BIN_DIR_NAME, 'pip')
-    exec_command(get_pip_install_command('azure-cli', path_to_pip, tmp_dir),
-                 env=dict(os.environ, AZURE_CLI_DISABLE_POST_INSTALL='1'))
-    modules_to_install = ['azure-cli-component', 'azure-cli-profile', 'azure-cli-storage',
-                          'azure-cli-vm', 'azure-cli-network', 'azure-cli-resource', 
-                          'azure-cli-role', 'azure-cli-taskhelp', 'azure-cli-feedback']
```
Comment (tjprescott) [inline scripts/install/install.py]: How do we control which modules are installed by default after this change?

File: src/azure-cli-core/azure/cli/core/application.py [hunk 3]
```diff
@@ -8,13 +8,13 @@
...
-from .parser import AzCliCommandParser
-from azure.cli._output import CommandResultItem
-import azure.cli.extensions
-import azure.cli._help as _help
-import azure.cli._logging as _logging
-from azure.cli._util import todict, CLIError
-from azure.cli._config import az_config
+from azure.cli.core.parser import AzCliCommandParser
+from azure.cli.core._output import CommandResultItem
+import azure.cli.core.extensions
```
Comment (tjprescott) [inline src/azure-cli-core/azure/cli/core/application.py]: Does it make sense for extensions to be part of the core... sort of by definition?

File: src/azure-cli-core/azure/cli/core/application.py [hunk 4]
```diff
@@ -65,7 +63,7 @@ def __init__(self, config=None):
         self.register(self.COMMAND_PARSER_PARSED, self._handle_builtin_arguments)
 
         # Let other extensions make their presence known
-        azure.cli.extensions.register_extensions(self)
+        azure.cli.core.extensions.register_extensions(self)
```
Comment (tjprescott) [inline src/azure-cli-core/azure/cli/core/application.py]: >register_extensions(self) [](start = 34, length = 25)

Okay, I can see this call being part of the core, but it seems the extensions themselves should not be part of the core... or they aren't really extensions. ;)

File: src/azure-cli-core/azure/cli/core/commands/template_create.py [hunk 5]
```diff
@@ -7,9 +7,9 @@
 import argparse
 import platform
 
-from azure.cli._util import CLIError
-from azure.cli.commands import register_cli_argument
-from azure.cli.commands.arm import (
+from azure.cli.core._util import CLIError
+from azure.cli.core.commands import register_cli_argument
+from azure.cli.core.commands.arm import (
```
Comment (tjprescott) [inline src/azure-cli-core/azure/cli/core/commands/template_create.py]: > azure.cli.core.commands.arm [](start = 4, length = 28)

Should ARM be part of core? (I'd say yes but I think it's worth asking the question).

File: src/azure-cli-core/azure/cli/core/tests/test_generic_update.py [hunk 6]
```diff
@@ -7,10 +7,10 @@
 import unittest
 import sys
 
-from azure.cli.application import APPLICATION, Application, Configuration
-from azure.cli.commands import CliArgumentType, register_cli_argument
-from azure.cli.commands.arm import cli_generic_update_command
-from azure.cli._util import CLIError
+from azure.cli.core.application import APPLICATION, Application, Configuration
+from azure.cli.core.commands import CliArgumentType, register_cli_argument
+from azure.cli.core.commands.arm import cli_generic_update_command
```
Comment (tjprescott) [inline src/azure-cli-core/azure/cli/core/tests/test_generic_update.py]: >.arm import cli_generic_update_command [](start = 28, length = 38)

I know this has nothing to do with your PR per se, but why is generic update part of arm.py? @johanste?

File: scripts/install/install.py (hunk 1 above)
Comment (derekbekoe) [inline scripts/install/install.py]: Yes as we don't need to have multiple virtual environments any longer.

File: scripts/install/install.py (hunk 2 above)
Comment (derekbekoe) [inline scripts/install/install.py]: In the setup.py file of azure-cli package

File: src/azure-cli-core/azure/cli/core/application.py (hunk 3 above)
Comment (derekbekoe) [inline src/azure-cli-core/azure/cli/core/application.py]: Yes, I have a note to move extensions out of core.
I'll do this as part of the work that defines exactly what goes into core.

File: src/azure-cli-core/azure/cli/core/application.py (hunk 4 above)
Comment (derekbekoe) [inline src/azure-cli-core/azure/cli/core/application.py]: agreed

File: src/azure-cli-core/azure/cli/core/utils/vcr_test_base.py [hunk 7]
```diff
@@ -23,11 +23,14 @@
 import jmespath
 from six import StringIO
 
-import azure.cli as cli
+# TODO Should not depend on azure.cli.main package here.
+# Will be ok if this test file is not part of azure.cli.core.utils
```
Comment (tjprescott) [inline src/azure-cli-core/azure/cli/core/utils/vcr_test_base.py]: Perhaps test should not be part of CLI core? Possibly part of extensions (assuming extensions are removed from the core)

File: src/command_modules/azure-cli-feedback/setup.py [hunk 8]
```diff
@@ -25,6 +25,7 @@
 
 DEPENDENCIES = [
     'applicationinsights',
```
Comment (tjprescott) [inline src/command_modules/azure-cli-feedback/setup.py]: > 'applicationinsights', [](start = 3, length = 23)

Is this an old hand authored SDK? The swagger for insights is currently in PR status being reviewed.

File: src/azure-cli-core/azure/cli/core/application.py (hunk 3 above)
Comment (tjprescott) [inline src/azure-cli-core/azure/cli/core/application.py]: Will that PR closely follow this one or should we create an issue to track it?

---
In reply to: [77924059](https://github.com/Azure/azure-cli/pull/828#discussion_r77924059) [](ancestors = 77924059)

Comment (derekbekoe) [inline src/azure-cli-core/azure/cli/core/application.py]: I'll create issues for the new work items

File: src/command_modules/azure-cli-feedback/setup.py (hunk 8 above)
Comment (derekbekoe) [inline src/command_modules/azure-cli-feedback/setup.py]: This is a different service

File: src/azure-cli-core/azure/cli/core/utils/vcr_test_base.py (hunk 7 above)
Comment (derekbekoe) [inline src/azure-cli-core/azure/cli/core/utils/vcr_test_base.py]: This was another note I had to not include test code as part of the product.

**Comment by tjprescott** [comment]
Looks good to me. Hopefully the follow on items follow soon. :)

---
In reply to: [245418417](https://github.com/Azure/azure-cli/pull/828#issuecomment-245418417) [](ancestors = 245418417)

**Comment by tjprescott** [comment]
:shipit:

File: scripts/command_modules/package_verify.py [hunk 9]
```diff
@@ -11,72 +11,56 @@
...
-print_heading('Installed CLI package.')
+# STEP 1:: Build the packages
+build_package(PATH_TO_AZURE_CLI, built_packages_dir)
+build_package(PATH_TO_AZURE_CLI_CORE, built_packages_dir)
+for name, fullpath in all_command_modules:
+    build_package(fullpath, built_packages_dir)
 
-print_heading('Installing command package(s)...')
-failed_module_names = []
+# STEP 3:: Install the CLI and dependencies
```
Comment (yugangw-msft) [inline scripts/command_modules/package_verify.py]: nit, this should be 'STEP 2'

File: src/azure-cli/setup.py [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+    'azure-cli-core',
+    'azure-cli-component',
+    'azure-cli-configure',
+    'azure-cli-feedback',
+    'azure-cli-network',
+    'azure-cli-profile',
+    'azure-cli-resource',
+    'azure-cli-role',
+    'azure-cli-storage',
+    'azure-cli-vm',
```
Comment (yugangw-msft) [inline src/azure-cli/setup.py]: Just a quick look, we don't have webapp. If it is intentionally, I will add later myself when on-boarding new commands

File: src/command_modules/azure-cli-webapp/setup.py [hunk 11]
```diff
@@ -25,6 +25,7 @@
 
 DEPENDENCIES = [
     'azure==2.0.0rc5',
+    'azure-cli-core',
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-webapp/setup.py]: support ask @lmazuel for the best practice, say, whether we need to put in a version range at the start.

**Comment by yugangw-msft** [comment]
:shipit:

File: src/azure-cli/setup.py (hunk 10 above)
Comment (derekbekoe) [inline src/azure-cli/setup.py]: Yes we are not including webapp by default right now.

**Pull request merged**